[iOS] Ensure Element.requestFullscreen reflects current hardware keyboard state during session

xingri opened the pull request:
#### b40d0b9064e69a6fb72def7fd4681a55546a7969
<pre>
[iOS] Ensure Element.requestFullscreen reflects current hardware keyboard state during session
<a href="https://bugs.webkit.org/show_bug.cgi?id=299626">https://bugs.webkit.org/show_bug.cgi?id=299626</a>

Reviewed by Abrar Rahman Protyasha.

This patch introdueces a ChromeClient implementation that allows synchronizing of hardware keyboard attachment and detachment status.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::requestFullscreen):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::addHardwareKeyboardAttachmentObserver):
(WebCore::Page::flushHardwareKeyboardAttachmentObservers):
(WebCore::Page::didUpdateHardwareKeyboardAttachment):
* Source/WebCore/page/Page.h:
(WebCore::Page::setHardwareKeyboardAttached): Deleted.
(WebCore::Page::hardwareKeyboardAttached const): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_toolbarsAreVisible):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::hardwareKeyboardAvailabilityChanged):

Canonical link: <a href="https://commits.webkit.org/303080@main">https://commits.webkit.org/303080@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/91f3fa05429f29eb4003d6e4a2f5b83d8458e10b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131199 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3529 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42163 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138642 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82907 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133069 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3527 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3370 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99960 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134145 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2524 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117446 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80660 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2441 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/148 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81891 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111035 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35564 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141141 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3274 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36091 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108482 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3319 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2926 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108425 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/27564 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2458 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113778 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56349 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3336 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32202 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3158 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66743 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3358 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3266 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

stwrt reviewed (changes requested):
Could you squash the two commits into one.

lilyspiniolas commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -67,6 +67,27 @@ class WebChromeClient final : public WebCore::ChromeClient {
+    {
+        m_callback = WTFMove(cb);
+        if (m_callback)
+            m_callback(m_isAttached);
```
```suggestion
        if (m_callback = WTFMove(cb))
            m_callback(m_isAttached);
```

lilyspiniolas commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5103,8 +5103,42 @@ void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredProm
+                        if (!protectedThis)
+                            return;
+
+                        auto* page = protectedThis->document().page();
```
```suggestion
                        RefPtr page = protectedThis->document().page();
```

lilyspiniolas commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5103,8 +5103,42 @@ void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredProm
+            if (!attached) {
+                protectedThis->protectedDocument()->postTask(
+                    WTF::Function<void()>([weakThis]() {
+                        auto protectedThis = weakThis.get();
```
```suggestion
                        RefPtr protectedThis = weakThis.get();
```

xingri replied:
Thanks @lilyspiniolas  for the review. Updated.

aprotyas commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -775,6 +775,16 @@ class ChromeClient {
+    virtual bool isHardwareKeyboardAttached() const { return false; }
+#else
+    virtual bool isHardwareKeyboardAttached() const { return true; }
+#endif
```
Why not just default this to `return true;`? You can change the return value in the concrete definition in `WebChromeClient`.

aprotyas replied:
the only concern would be WebKitLegacy but fullscreen keyboard lock will be unconditionally off there anyway.

aprotyas commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -579,6 +599,13 @@ class WebChromeClient final : public WebCore::ChromeClient {
+    bool m_isAttached { false };
+#else
+    bool m_isAttached { true };
+#endif
```
Two things:

1. `m_isAttached` is quite general, it's hard to understand that this is about hardware keyboard attachment.
2. Why do we need to store the `m_isAttached` flag regardless? I think you can totally read this from `WebPage::hardwareKeyboardIsAttached()`. We should try our best not to duplicate state here because of the asynchronous nature in which keyboard attachment state propagates.

aprotyas commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -67,6 +67,26 @@ class WebChromeClient final : public WebCore::ChromeClient {
+        m_isAttached = attached;
+        if (m_callback)
+            m_callback(attached);
+    }
```
Related to my other comment https://github.com/WebKit/WebKit/pull/52418/files#r2470706271:

Essentially what I'm suggesting is

```cpp
bool WebChromeClient::isHardwareKeyboardAttached() const override
{
    RefPtr page = m_page.get();
    return page && page->hardwareKeyboardIsAttached();
}
```

aprotyas replied:
Can you define these methods in the cpp file btw? Some of them are not inline friendly.

aprotyas commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -67,6 +67,26 @@ class WebChromeClient final : public WebCore::ChromeClient {
 
+    bool isHardwareKeyboardAttached() const override { return m_isAttached; }
+
+    void startHardwareKeyboardAttachmentNotifications(HardwareKeyboardAttachmentCallback&& cb) override
```
The style guide suggests going with full names, unless the abbreviation is canonical
```suggestion
    void startHardwareKeyboardAttachmentNotifications(HardwareKeyboardAttachmentCallback&& callback) override
```

aprotyas commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -67,6 +67,26 @@ class WebChromeClient final : public WebCore::ChromeClient {
+
+    void startHardwareKeyboardAttachmentNotifications(HardwareKeyboardAttachmentCallback&& cb) override
+    {
+        if ((m_callback = WTFMove(cb)))
```
Extra parentheses
```suggestion
        if (m_callback = WTFMove(callback))
```

aprotyas commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -67,6 +67,26 @@ class WebChromeClient final : public WebCore::ChromeClient {
+    void startHardwareKeyboardAttachmentNotifications(HardwareKeyboardAttachmentCallback&& cb) override
+    {
+        if ((m_callback = WTFMove(cb)))
+            m_callback(m_isAttached);
```
Since I'm suggesting to get rid of `m_isAttached` and just reading the value from `WebPage`, just delegate logic to the getter:


```suggestion
            m_callback(isHardwareKeyboardAttached());
```

aprotyas commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -579,6 +599,13 @@ class WebChromeClient final : public WebCore::ChromeClient {
+#else
+    bool m_isAttached { true };
+#endif
+    HardwareKeyboardAttachmentCallback m_callback;
```
`m_callback` is generic, maybe we want to say `m_hardwareKeyboardAttachmentCallback`?

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -5710,6 +5710,9 @@ static bool hasAccessoryStylusPointingDevice()
     m_keyboardIsAttached = state.isAttached;
     setHardwareKeyboardState(state);
 
+    // Notify Hardware Keyboard attachment state WebCore via ChromeClient
```
I think we can remove this comment since the code is self-explanatory here.

aprotyas commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5107,8 +5107,43 @@ void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredProm
+                            doc->fullscreen().fullyExitFullscreen();
+                    })
+                );
+            }
```
nit: I would do an early return
```suggestion
            if (attached)
                return;

            protectedThis->protectedDocument()->postTask(
                WTF::Function<void()>([weakThis]() {
                    RefPtr protectedThis = weakThis.get();
                    if (!protectedThis)
                        return;

                    RefPtr page = protectedThis->document().page();
                    if (!page)
                        return;

                    auto& chromeClient = page->chrome().client();
                    chromeClient.stopHardwareKeyboardAttachmentNotifications();

                    Ref<Document> doc = protectedThis->protectedDocument().get();
                    if (doc->fullscreen().fullscreenElement())
                        doc->fullscreen().fullyExitFullscreen();
                })
            );
```

aprotyas commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5107,8 +5107,43 @@ void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredProm
+            // Exit the fullscreen API when the hardware keyboard is detached.
+            if (!attached) {
+                protectedThis->protectedDocument()->postTask(
+                    WTF::Function<void()>([weakThis]() {
```
Do you need to explicitly create `WTF::Function`? I feel like any lambda you pass in should work? You could try this:

```
protectedThis->protectedDocument()->postTask([weakThis] {
    ...
});
```

aprotyas replied:
if you have to explicitly specify `WTF::Function`, can you remove the `WTF` prefix? Typically we do not use that namespace prefix in the project.

aprotyas replied:
Also, last nit, you can skip empty lambda arg lists in C++23
```suggestion
                    WTF::Function<void()>([weakThis] {
```

aprotyas commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5107,8 +5107,43 @@ void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredProm
+                        auto& chromeClient = page->chrome().client();
+                        chromeClient.stopHardwareKeyboardAttachmentNotifications();
+
+                        Ref<Document> doc = protectedThis->protectedDocument().get();
```
safer-cpp style is to store a Ref from the non-protected getters:

```suggestion
                        Ref document = protectedThis->document();
```

aprotyas commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5107,8 +5107,43 @@ void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredProm
+
+        // Registers a callback to exit fullscreen mode
+        client.startHardwareKeyboardAttachmentNotifications([weakThis = WeakPtr { *this }](bool attached) {
+            auto protectedThis = weakThis.get();
```
```suggestion
            RefPtr protectedThis = weakThis.get();
```

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -5710,6 +5710,9 @@ static bool hasAccessoryStylusPointingDevice()
     setHardwareKeyboardState(state);
 
+    // Notify Hardware Keyboard attachment state WebCore via ChromeClient
+    chromeClient().didUpdateHardwareKeyboardAttachment(m_keyboardIsAttached);
```
I think we need to rethink about the data flow here. It's weird to call back into ChromeClient from WebPage

aprotyas reviewed (commented):
Thanks for the patch, this is a first pass.

I think we need to re-think the data flow here. The WebCore::Page getter (`hardwareKeyboardAttached()`) is fine and enough. I think what you want is this:

1. In `WebCore::Element`, ask for keyboard attachment state from `WebCore::Page`. `WebCore::Page::hardwareKeyboardAttached()` can query `[Web]ChromeClient`, which queries `WebKit::Page` for the ground truth.

2. Register the hardware keyboard attachment callbacks with `WebCore::Page`. When the web process knows that attachment state changes, it will call `WebCore::Page::didUpdateHardwareKeyboardAttachment()` from `WebKit::WebPage::hardwareKeyboardAvailabilityChanged()`.

Basically, ChromeClient should not be holding on to state, and WebKit::WebPage shouldn't be calling into ChromeClient, rather it should call directly into WebCore::Page

aprotyas commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -781,6 +781,13 @@ class ChromeClient {
+    using HardwareKeyboardAttachmentObserver = Function<void(bool)>;
+    virtual void addHardwareKeyboardAttachmentObserver(HardwareKeyboardAttachmentObserver&&) { }
+
+    virtual bool isHardwareKeyboardAttached() const { return true; }
```
I think it's better without the `is` prefix
```suggestion
    virtual bool hardwareKeyboardAttached() const { return true; }
```

aprotyas commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -781,6 +781,13 @@ class ChromeClient {
+
+    virtual bool isHardwareKeyboardAttached() const { return true; }
+
+    virtual bool isMockChromeClient() const { return false; }
```
This is probably unnecessary now given no testing support in this iteration?

aprotyas commented on Source/WebCore/page/Page.h:
```diff
@@ -1390,8 +1390,31 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
+            return;
+
+        m_hardwareKeyboardAttached = attached;
+    }
```
You should probably move these definitions to Page.cpp

aprotyas commented on Source/WebCore/page/Page.h:
```diff
@@ -1390,8 +1390,31 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
+            return;
+
+        m_hardwareKeyboardAttached = attached;
+    }
```
I would expect `flushHardwareKeyboardAttachmentObservers()` to be an implementation detail private to `WebCore::Page` that gets called automatically whenever we alter `m_hardwareKeyboardAttached` in `didUpdateHardwareKeyboardAttachment()`.

It seems odd that `WebPage` has the duty to call it, and even then we don't call it unconditionally?

aprotyas replied:
More concretely, I would expect:

```cpp
    void didUpdateHardwareKeyboardAttachment(bool attached)
    {
        if (m_hardwareKeyboardAttached == attached)
            return;

        m_hardwareKeyboardAttached = attached;
        flushHardwareKeyboardAttachmentObservers();
    }
```

... plus, that lets us make `flushHardwareKeyboardAttachmentObservers()` a private method.

aprotyas commented on Source/WebCore/page/Page.h:
```diff
@@ -1390,8 +1390,31 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
+        m_hardwareKeyboardAttached = attached;
+    }
+
     bool hardwareKeyboardAttached() const { return m_hardwareKeyboardAttached; }
```
I understand you are storing this state internally in `WebCore::Page` to be able to carefully flush the observers, but no one should be able to read it. The ground truth should be read from `WebPage`, so I would suggest to remove this `hardwareKeyboardAttached()` getter from this class.

aprotyas commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5066,8 +5066,39 @@ void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredProm
+        auto& client = document().page()->chrome().client();
+
+        // Registers a callback to exit fullscreen mode
+        client.addHardwareKeyboardAttachmentObserver([weakThis = WeakPtr { *this }](bool attached) {
```
Wrong direction with this one, you don't need to teach `Client`/`WebPage` about the observers, you can add it directly to `WebCore::Page`.

i.e. `document().page()->addHardwareKeyboardAttachmentObserver(...)`

aprotyas replied:
and then you can remove `addHardwareKeyboardAttachmentObserver()` from `ChromeClient`/`WebChromeClient`/`WebPage`

aprotyas commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5066,8 +5066,39 @@ void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredProm
+                if (!page)
+                    return;
+
+                Ref<Document> doc = protectedThis->protectedDocument().get();
```
```suggestion
                Ref document = protectedThis->protectedDocument();
```

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -5740,6 +5740,9 @@ static bool hasAccessoryStylusPointingDevice()
     setHardwareKeyboardState(state);
 
+    didUpdateHardwareKeyboardAttachment(m_keyboardIsAttached);
+    flushHardwareKeyboardAttachmentObservers();
```
This seems like a logic error, you should stuff the `flushHardwareKeyboardAttachmentObservers()` call inside `didUpdateHardwareKeyboardAttachment()`.

aprotyas replied:
Also, you should directly call `page->didUpdateHardwareKeyboardAttachment(...)` from here. (like you do in `WebPage::WebPage()`)

aprotyas commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -2394,4 +2394,22 @@ void WebChromeClient::showCaptionDisplaySettings(CompletionHandler<void(bool)>&&
+    if (!page)
+        return false;
+
+    return page->hardwareKeyboardAttached();
```
`WebPage::hardwareKeyboardIsAttached()` already exists, you should call that instead of this new method.

aprotyas commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -2394,4 +2394,22 @@ void WebChromeClient::showCaptionDisplaySettings(CompletionHandler<void(bool)>&&
+        return;
+
+    page->addHardwareKeyboardAttachmentObserver(WTFMove(observer));
+}
```
Not necessary, read my other comment about calling `WebCore::Page::addHardwareKeyboardAttachmentObserver()` directly.

aprotyas commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -10291,6 +10292,36 @@ void WebPage::hideCaptionDisplaySettingsPreview(HTMLMediaElementIdentifier ident
+{
+    RefPtr page { corePage() };
+    return page ? page->hardwareKeyboardAttached() : false;
+}
```
Please remove this, we should not be consulting `WebCore::Page`. Plus, `hardwareKeyboardIsAttached()` already exists.

aprotyas commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -10291,6 +10292,36 @@ void WebPage::hideCaptionDisplaySettingsPreview(HTMLMediaElementIdentifier ident
+    if (!page)
+        return;
+    page->flushHardwareKeyboardAttachmentObservers();
+}
```
Please remove these.

aprotyas commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -1204,6 +1206,11 @@ class WebPage final : public API::ObjectImpl<API::Object::Type::BundlePage>, pub
+    bool hardwareKeyboardAttached() const;
+    void addHardwareKeyboardAttachmentObserver(WebCore::HardwareKeyboardAttachmentObserver&&);
+    void flushHardwareKeyboardAttachmentObservers();
+    void didUpdateHardwareKeyboardAttachment(bool);
```
Please remove.

aprotyas commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -378,6 +378,8 @@ using SessionID = WTF::UUID;
 using TextSuggestionID = WTF::UUID;
 } // namespace WritingTools
 
+using HardwareKeyboardAttachmentObserver = Function<void(bool)>;
```
Please remove.

aprotyas reviewed (commented):
This is looking much better, thank you for iterating.

We have a bunch of unnecessary plumbing/state right now that we can cull. Concretely:

1. Teach either WebCore::Page or WebKit::WebPage about hardware keyboard attachment observers. My vote goes to the former (less plumbing that way).

    1. Currently you are doing an awkward dance of Element -> ChromeClient -> WebPage -> Page to add the observers, when you could simply do Element -> Page.
2. Whenever hardware keyboard attachment changes in WebPage, just call WebCore::Page::didUpdateHardwareKeyboardAttachment(), and make that method execute the observers under the hood.

aprotyas commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5066,8 +5066,34 @@ void Element::requestFullscreen(FullscreenOptions&& options, RefPtr<DeferredProm
+#if PLATFORM(IOS_FAMILY)
+        // Registers a callback to exit fullscreen mode
+        document().page()->addHardwareKeyboardAttachmentObserver([weakThis = WeakPtr { *this }](bool attached) {
+
```
Nit: remove this extra new line

xingri committed:
[iOS] Ensure Element.requestFullscreen reflects current hardware keyboard state during session
https://bugs.webkit.org/show_bug.cgi?id=299626

Reviewed by Abrar Rahman Protyasha.

This patch introdueces a ChromeClient implementation that allows synchronizing of hardware keyboard attachment and detachment status.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::requestFullscreen):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::addHardwareKeyboardAttachmentObserver):
(WebCore::Page::flushHardwareKeyboardAttachmentObservers):
(WebCore::Page::didUpdateHardwareKeyboardAttachment):
* Source/WebCore/page/Page.h:
(WebCore::Page::setHardwareKeyboardAttached): Deleted.
(WebCore::Page::hardwareKeyboardAttached const): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_toolbarsAreVisible):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::hardwareKeyboardAvailabilityChanged):

Canonical link: https://commits.webkit.org/303080@main
